man/xdd.1
.\" Text automatically generated by txt2man
.TH untitled  "19 June 2023" "" ""
.SH NAME
\fBxdd \fP- No short description
.SH SYNOPSIS
.nf
.fam C
\fBxdd\fP <command> [<args>]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
No long description
.SH COMMANDS
.TP
.B
\fBxdd-sync\fP(1)
Copy \fB--data-nbytes\fP from \fB--data-input\fP to \fB--data\fP-\fB-output\fP
.TP
.B
\fBxdd-async\fP(1)
Copy \fB--data-nbytes\fP from \fB--data-input\fP to \fB--data\fP-\fB-output\fP
.RE
.PP

.SH OPTIONS
\fB--help\fP
Print the synopsis and exit
.SH EXAMPLES
Read the man page for each <command> or consult the command-line \fB--help\fP:
.PP
.nf
.fam C
    $ xdd <command> --help

.fam T
.fi
.SH SEE ALSO
Full documentation at: <https://xnvme.io/>
.SH AUTHOR
Written by Simon A. F. Lund <simon.lund@samsung.com> on behalf of Samsung


CHANGELOG.rst
.. SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
..
.. SPDX-License-Identifier: BSD-3-Clause

Changelog
=========

The repository is tagged using semantic versioning, e.g. ``v1.2.3``.

The ``next`` branch contains a preview of the upcoming release. The history on
``next`` is subject to change. Upon release, ``next`` is merged with ``main``
and  and tagged ``vX.Y.Z``.

Changes are described in this file in a section matching the version tag.

Known Issues
------------

See the file named ``ISSUES`` in the root of the repository.

v0.7.0
------

This release got a whole lot bigger than it ought to. Aiming for more
frequent releases in the future. A bunch of changes to API,
implementation, and even the license.

Regarding the license. For compatibility with GPL-licensed software.
That is, consumption of xNVMe from GPL-licences software, then xNVMe is
changed from APACHE to BSD-3-Clause. To better determine the license,
the project has aligned with the golden practice of REUSE-compliance
(https://reuse.software/).

* API
  * Refactored to provide a single "entry-point" header
  * Headers no longer include other headers (except for the entry-point)
  * Renamed the command-line api from xnvmec to xnvme_cli

* API / Command-Sets
  * Support for FDP and DSM
  * Support for Key-Value SSDs

* Backends
  * The POSIX backend is replaced by CBI -- Common Backend Implementation
  * Instead of "POSIX" then a handful of commonly useful implementations
    are provided, being "strictly" POSIX was an impractical restriction,
    such as lacking support for features which are provided on modern
    operating systems but not defined by POSIX.

* Backends: FreeBSD
  * Support for kqueue based aio

* Backends: SPDK
  * Now has options to control Admin and I/O command timeout
  * Removed patches no longer needed (available upstream)

* Backends: vfio
  * Removed as subproject and instead links "properly" with the
    system-provided libvfn

* tools
  * switched fio from "external" to "internal" / "upstream" fio engine

For additional details then please have a look at the commit-history,
the v0.7.0 related issue on GitHUB and PR.

v0.6.0
------

A handful of improvements for Windows, additional steps toward removing
third-party vendoring / bundling, addition of experimental tunable knobs for
io_uring/SQPOLL, and several improvements to testing and CI infra.

* API
  - Removed SLIST from API, although "sys/queue.h" are commonly available on
    Linux/FreeBSD, then they are not part of toolchain on Windows.

* Third-party
  - Bumped SPDK to v22.09, and with that removed mutliple out-of-tree patches
    for DPDK.

* CLI
  - The xNVMe command-line library (libxnvmec) and all the cli-tools using it
    are refactored to use common sets of command-line arguments. Along with
    this came a consistent set of CLI-arguments for admin/sync/async.

* Backends
  - ramdisk: The ramdisk got support for write-zeroes, iovec payloads and added
    to CI testing.
  - linux: support for a buffer-allocator using HUGEPAGES and tunable knobs for
    controlling the behavior of io_uring SQPOLL via environment variables.
  - windows: support for the experimental IORING Windows SPDK API and support
    for block devices (SCSI and SATA).
  - spdk: when controllers are re-used for device-handles, events are
    processed as a means to check whether the controller is still "alive"

* CI
  - scan-build now runs for each PR
  - basic tests are now running post-building testing using the RAMDISK
  - Testing of fabrics with TCP transport is now part of the setup

v0.5.0
------

A bit of expansion in the application of xNVMe with support for macOS, a
ramdisk backend, revival of Python language bindings and a refresh of the docs
on NVMe-over-Fabrics.

* API
  - Removed helpers for SGL
  - Add 'subnqn' to 'xnvme_ident', useful for NVMe-oF
  - Add 'hostnqn' to 'opts', useful for NVMe-oF
  - Add support directive-receive and write-with-directives
  - Spec adjusted for NVMe 2.0, still more work needed in this area

* Third-party
  - Bumped fio to v3.32
  - Bumped SPDK to v22.05
  - Removed liburing, now relies on on-system library instead of
    vendoring/bundling, documentation is updated to assist with library
    installation

* Tooling
  - Re-working testing using cijoe 0.9+, that is, switching to CIJOE/pytest for
    testing an CIJOE workflows for instrumentation
  - liburing is no longer bundled with xNVMe, that is, xNVMe now links with
    liburing as discovered on the system via pkg-config. This is done to avoid
    symbol collisions for applications linking or loading liburing and xNVMe.

* be:linux:async:libaio
  - When 'opts.poll_io' is set then poke() will return immmediatly and now wait
    for completions. This allows the use of trading CPU for more IOPS and lower
    per command latency.

* Additional user-space NVMe driver support via libvfn
  - Added 'be:vfio' providing another user-space driver via libvfn

* Preliminary support for macOS
  - Initial implementation using the "core" I/O mechanisms of sync-io, async
    emulation and the threadpool
  - Does enumeration of NVMe devices through the limited interface provided for
    user-space by the macOS kernel
  - Utilizes what is avaiable for admin-command submision

* Prelimiary support for a "ramdisk" device
  - be:ramdisk: added a backend mimicing the behavior of an NVMe NVM namespace
  - Intended as a test-vehicle providing a "device" without requiring anything
    but the consumption of main memory of the system
  - I/O is "stored" using main-memory

* Revival of the xNVMe Python language bindings
  - A re-introduction of the Python bindings, these are now generated and thus
    provide access to the entire xNVMe C API
  - They are added to the testing infrastructure ensuring that they are aligned
    with the library
  - In addition to simple ctypes bindings, then cython headers and bindings
    based on Cython are provided

* Documentation

tools/xdd.c
// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
//
// SPDX-License-Identifier: BSD-3-Clause

#include <errno.h>
#include <libxnvme.h>

#define IOSIZE_DEF 4096
#define QDEPTH_MAX 256
#define QDEPTH_DEF 16
#define START_OFFSET_DEF 0

struct cb_args {
	uint64_t *nerrors;
	uint64_t ncompletions;
	uint64_t nsubmissions;
	uint64_t start_offset;

	struct xnvme_cmd_ctx dst_ctx;
	char *buf;
};

static void
cb_func(struct xnvme_cmd_ctx *ctx, void *cb_arg)
{
	struct cb_args *work = cb_arg;

	work->ncompletions += 1;

	if (xnvme_cmd_ctx_cpl_status(ctx)) {
		xnvme_cli_perr("cb_func()", errno);
		xnvme_cmd_ctx_pr(ctx, XNVME_PR_DEF);
		*work->nerrors += 1;
	} else {
		ssize_t res;

		res = xnvme_file_pwrite(&work->dst_ctx, work->buf, ctx->cpl.result,
					ctx->cmd.nvm.slba - work->start_offset);
		if (res || xnvme_cmd_ctx_cpl_status(&work->dst_ctx)) {
			xnvme_cli_perr("xnvme_file_pwrite(dst)", res);
			xnvme_cmd_ctx_pr(&work->dst_ctx, XNVME_PR_DEF);
			*work->nerrors += 1;
		}
	}

	xnvme_queue_put_cmd_ctx(ctx->async.queue, ctx);
}

int
copy_async(struct xnvme_cli *cli)
{
	const char *src_uri, *dst_uri;
	struct xnvme_dev *src_dev, *dst_dev;
	struct xnvme_opts src_opts = {.rdonly = 1, .direct = cli->args.direct};
	struct xnvme_opts dst_opts = {.wronly = 1, .create = 1, .direct = cli->args.direct};
	size_t buf_nbytes, tbytes, iosize, start_offset;
	char *buf = NULL;

	struct cb_args cb_args[QDEPTH_MAX] = {0};
	uint64_t nerrors = 0, ncompletions = 0, nsubmissions = 0;
	struct xnvme_queue *queue = NULL;
	uint32_t qdepth;
	int err;

	src_uri = cli->args.data_input;
	dst_uri = cli->args.data_output;
	tbytes = cli->args.data_nbytes;

	iosize = cli->given[XNVME_CLI_OPT_IOSIZE] ? cli->args.iosize : IOSIZE_DEF;
	if (!iosize) {
		err = -EINVAL;
		xnvme_cli_perr("iosize can't be set to 0", err);
		return err;
	}

	qdepth = cli->given[XNVME_CLI_OPT_QDEPTH] ? cli->args.qdepth : QDEPTH_DEF;
	if (!qdepth) {
		err = -EINVAL;
		xnvme_cli_perr("qdepth can't be set to 0", err);
		return err;
	}
	start_offset = cli->given[XNVME_CLI_OPT_OFFSET] ? cli->args.offset : START_OFFSET_DEF;

	src_dev = xnvme_file_open(src_uri, &src_opts);
	if (src_dev == NULL) {
		xnvme_cli_perr("xnvme_file_open(src)", errno);
		return errno;
	}
	dst_dev = xnvme_file_open(dst_uri, &dst_opts);
	if (dst_dev == NULL) {
		xnvme_cli_perr("xnvme_file_open(dst)", errno);
		xnvme_file_close(src_dev);
		return errno;
	}

	buf_nbytes = iosize * qdepth;
	buf = xnvme_buf_alloc(dst_dev, buf_nbytes);
	if (!buf) {
		xnvme_cli_perr("xnvme_buf_alloc()", errno);
		goto exit;
	}
	xnvme_buf_fill(buf, buf_nbytes, "zero");

	err = xnvme_queue_init(src_dev, qdepth, 0, &queue);
	if (err) {
		xnvme_cli_perr("xnvme_queue_init()", err);
		goto exit;
	}
	xnvme_queue_set_cb(queue, cb_func, &cb_args);

	xnvme_cli_pinf("copy-async: "
		       "{src: %s, dst: %s, tbytes: %zu, buf_nbytes: %zu, iosize: %zu, qdepth: %u, "
		       "start_offset: %zu}",
		       src_uri, dst_uri, tbytes, buf_nbytes, iosize, qdepth, start_offset);

	xnvme_cli_timer_start(cli);

	for (uint32_t i = 0; i < qdepth; ++i) {
		struct cb_args *work = &cb_args[i];

		work->nerrors = &nerrors;
		work->buf = buf + (i * iosize);
		work->dst_ctx = xnvme_file_get_cmd_ctx(dst_dev);
		work->start_offset = start_offset;
	}
	for (size_t ofz = start_offset; (ofz - start_offset < tbytes) && !nerrors;) {
		struct cb_args *work = &cb_args[((ofz - start_offset) / iosize) % qdepth];
		struct xnvme_cmd_ctx *src_ctx = xnvme_cmd_ctx_from_queue(queue);
		size_t nbytes = XNVME_MIN_U64(iosize, tbytes - (ofz - start_offset));
		ssize_t res;

		src_ctx->async.cb_arg = work;

submit:
		res = xnvme_file_pread(src_ctx, work->buf, nbytes, ofz);
		switch (res) {
		case 0:
			work->nsubmissions += 1;
			goto next;

		case -EBUSY:
		case -EAGAIN:
			xnvme_queue_poke(queue, 0);
			goto submit;

		default:
			xnvme_cli_perr("submission-error", err);
			xnvme_queue_put_cmd_ctx(queue, src_ctx);
			goto exit;
		}

next:
		ofz += iosize;
	}

	err = xnvme_queue_drain(queue);
	if (err < 0) {
		xnvme_cli_perr("xnvme_queue_drain()", err);
		goto exit;
	}

	xnvme_cli_timer_stop(cli);
	xnvme_cli_timer_bw_pr(cli, "wall-clock", tbytes);

exit:
	for (uint32_t i = 0; i < qdepth; ++i) {
		struct cb_args *work = &cb_args[i];

		nsubmissions += work->nsubmissions;
		ncompletions += work->ncompletions;
	}
	xnvme_cli_pinf("cb_args: {nsubmissions: %zu, ncompletions: %zu, nerrors: %zu}",
		       nsubmissions, ncompletions, nerrors);

	if (queue) {
		int err_exit = xnvme_queue_term(queue);
		if (err_exit) {
			xnvme_cli_perr("xnvme_queue_term()", err_exit);
		}
	}

	xnvme_buf_free(dst_dev, buf);
	xnvme_file_close(src_dev);
	xnvme_file_close(dst_dev);
	return 0;
}

int
copy_sync(struct xnvme_cli *cli)
{
	const char *src_uri, *dst_uri;
	struct xnvme_dev *src_dev, *dst_dev;
	struct xnvme_opts src_opts = {.rdonly = 1, .direct = cli->args.direct};
	struct xnvme_opts dst_opts = {.wronly = 1, .create = 1, .direct = cli->args.direct};
	size_t buf_nbytes, tbytes, iosize, start_offset;
	char *buf = NULL;

	src_uri = cli->args.data_input;
	dst_uri = cli->args.data_output;
	tbytes = cli->args.data_nbytes;

	iosize = cli->given[XNVME_CLI_OPT_IOSIZE] ? cli->args.iosize : IOSIZE_DEF;
	start_offset = cli->given[XNVME_CLI_OPT_OFFSET] ? cli->args.offset : START_OFFSET_DEF;

	src_dev = xnvme_file_open(src_uri, &src_opts);
	if (!src_dev) {
		xnvme_cli_perr("xnvme_file_open(src)", errno);
		return errno;
	}
	dst_dev = xnvme_file_open(dst_uri, &dst_opts);
	if (!dst_dev) {
		xnvme_cli_perr("xnvme_file_open(dst)", errno);
		xnvme_file_close(src_dev);
		return errno;
	}

	buf_nbytes = iosize;
	buf = xnvme_buf_alloc(dst_dev, buf_nbytes);
	if (!buf) {
		xnvme_cli_perr("xnvme_buf_alloc()", errno);
		goto exit;
	}
	xnvme_buf_fill(buf, buf_nbytes, "zero");

	xnvme_cli_pinf("copy-sync: {src: %s, dst: %s, tbytes: %zu, buf_nbytes: %zu, iosize: %zu, "
		       "start_offset: %zu}",
		       src_uri, dst_uri, tbytes, buf_nbytes, iosize, start_offset);

	xnvme_cli_timer_start(cli);

	for (size_t ofz = start_offset; ofz - start_offset < tbytes; ofz += iosize) {
		struct xnvme_cmd_ctx src_ctx = xnvme_file_get_cmd_ctx(src_dev);
		struct xnvme_cmd_ctx dst_ctx = xnvme_file_get_cmd_ctx(dst_dev);
		size_t nbytes = XNVME_MIN_U64(iosize, tbytes - (ofz - start_offset));
		ssize_t res;

		res = xnvme_file_pread(&src_ctx, buf, nbytes, ofz);
		if (res || xnvme_cmd_ctx_cpl_status(&src_ctx)) {
			xnvme_cli_perr("xnvme_file_pread(src)", res);
			xnvme_cmd_ctx_pr(&src_ctx, XNVME_PR_DEF);
			goto exit;
		}

		res = xnvme_file_pwrite(&dst_ctx, buf, nbytes, ofz - start_offset);
		if (res || xnvme_cmd_ctx_cpl_status(&dst_ctx)) {
			xnvme_cli_perr("xnvme_file_pwrite(dst)", res);
			xnvme_cmd_ctx_pr(&dst_ctx, XNVME_PR_DEF);
			goto exit;
		}
	}

	xnvme_cli_timer_stop(cli);
	xnvme_cli_timer_bw_pr(cli, "wall-clock", tbytes);

exit:
	xnvme_buf_free(dst_dev, buf);
	xnvme_file_close(src_dev);
	xnvme_file_close(dst_dev);
	return 0;
}

static struct xnvme_cli_sub g_subs[] = {
	{
		"sync",
		"Copy --data-nbytes from --data-input to --data--output",
		"Copy --data-nbytes from --data-input to --data--output",
		copy_sync,
		{
			{XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},
			{XNVME_CLI_OPT_DATA_INPUT, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_DATA_OUTPUT, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_DATA_NBYTES, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_IOSIZE, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_DIRECT, XNVME_CLI_LFLG},
			{XNVME_CLI_OPT_OFFSET, XNVME_CLI_LOPT},
		},
	},
	{
		"async",
		"Copy --data-nbytes from --data-input to --data--output",
		"Copy --data-nbytes from --data-input to --data--output",
		copy_async,
		{
			{XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},
			{XNVME_CLI_OPT_DATA_INPUT, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_DATA_OUTPUT, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_DATA_NBYTES, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_IOSIZE, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_QDEPTH, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_DIRECT, XNVME_CLI_LFLG},
			{XNVME_CLI_OPT_OFFSET, XNVME_CLI_LOPT},
		},
	},
};

static struct xnvme_cli g_cli = {
	.title = "xNVMe dd - Copy bytes from input to output",
	.descr_short = "Copy bytes from input to output",
	.descr_long = "",
	.subs = g_subs,
	.nsubs = sizeof g_subs / sizeof(*g_subs),
};

int
main(int argc, char **argv)
{
	return xnvme_cli_run(&g_cli, argc, argv, XNVME_CLI_INIT_NONE);
}
